# SmartHire AI/ML Project

## Data Validation

### Objective

The purpose of this notebook is to validate the processed datasets generated by the preprocessing pipeline before performing Exploratory Data Analysis (EDA) and Machine Learning.

This notebook verifies:

- Dataset availability
- Dataset dimensions
- Column names
- Data types
- Missing values
- Duplicate records
- Sample records

Only after successful validation will the datasets be used for EDA and model development.

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

In [2]:
PROJECT_ROOT = Path("..")
PROCESSED_DATA = PROJECT_ROOT / "data" / "processed"
INTERIM_DATA = PROJECT_ROOT / "data" / "interim"

print("Processed Data Folder:", PROCESSED_DATA)
print("Interim Data Folder:", INTERIM_DATA)

Processed Data Folder: ..\data\processed
Interim Data Folder: ..\data\interim


In [3]:
resume_df = pd.read_csv(PROCESSED_DATA/"resumes_clean.csv")
jobs_df = pd.read_csv(PROCESSED_DATA/"jobs_clean.csv")
corpus_df = pd.read_csv(INTERIM_DATA/"job_corpus.csv")

print("Datasets loaded successfully!")

Datasets loaded successfully!


In [4]:
print(f"Resume Dataset: {resume_df.shape}")
print(f"Jobs Dataset: {jobs_df.shape}")
print(f"Job Corpus: {corpus_df.shape}")

Resume Dataset: (962, 2)
Jobs Dataset: (136759, 8)
Job Corpus: (136759, 7)


In [7]:
print(resume_df.columns.tolist())

print(jobs_df.columns.tolist())

print(corpus_df.columns.tolist())

['Category', 'text']
['title', 'company', 'location', 'skills', 'description', 'experience', 'source', 'text']
['title', 'company', 'location', 'skills', 'description', 'experience', 'source']


## Dataset Validation Summary

In this section, we verify that all processed datasets were loaded correctly before performing detailed validation.

The following checks have been completed:

- Dataset loading
- Dataset dimensions
- Column validation

The next steps will validate data quality by checking:
- Data types
- Missing values
- Duplicate records
- Sample observations

In [31]:
# Dataset Information

resume_df.info()

print("#"*50)

jobs_df.info()

print("#"*50)

corpus_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 962 entries, 0 to 961
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   Category  962 non-null    str  
 1   text      962 non-null    str  
dtypes: str(2)
memory usage: 2.7 MB
##################################################
<class 'pandas.DataFrame'>
RangeIndex: 136759 entries, 0 to 136758
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype
---  ------       --------------   -----
 0   title        136759 non-null  str  
 1   company      135111 non-null  str  
 2   location     136261 non-null  str  
 3   skills       134512 non-null  str  
 4   description  136748 non-null  str  
 5   experience   109417 non-null  str  
 6   source       136759 non-null  str  
 7   text         136759 non-null  str  
dtypes: str(8)
memory usage: 934.7 MB
##################################################
<class 'pandas.DataFrame'>
RangeIndex: 136759 entries, 0 to 136758
Data co

In [19]:
# Missing Values

print("\nResume Dataset")
print("#"*30)
print(resume_df.isnull().sum())

print("\nJobs Dataset")
print("#"*30)
print(jobs_df.isnull().sum())

print("\nJob Corpus")
print("#"*30)
display(corpus_df.isnull().sum())


Resume Dataset
##############################
Category    0
text        0
dtype: int64

Jobs Dataset
##############################
title              0
company         1648
location         498
skills          2247
description       11
experience     27342
source             0
text               0
dtype: int64

Job Corpus
##############################


title              0
company         1648
location         498
skills          2247
description       11
experience     27342
source             0
dtype: int64

In [21]:
# Duplicate Records

print(f"Resume Dataset : {resume_df.duplicated().sum()}")
print(f"Jobs Dataset   : {jobs_df.duplicated().sum()}")
print(f"Job Corpus     : {corpus_df.duplicated().sum()}")

Resume Dataset : 796
Jobs Dataset   : 0
Job Corpus     : 0


In [25]:
# First Five Records

print("Resume Dataset")
print("#"*30)
display(resume_df.head())

Resume Dataset
##############################


,Category,text
0,Data Science,skills programming languages python pandas num...
1,Data Science,education details may 2013 to may 2017 b.e uit...
2,Data Science,areas of interest deep learning control system...
3,Data Science,skills r python sap hana tableau sap hana sql ...
4,Data Science,education details mca ymcaust faridabad haryan...


In [26]:
print("Jobs Dataset")
print("#"*30)
display(jobs_df.head())

Jobs Dataset
##############################


,title,company,location,skills,description,experience,source,text
0,Walkin Data Entry Operator (night Shift),MM Media Pvt Ltd,Chennai,ITES,Job Description Send me Jobs like this Quali...,0 - 1 yrs,naukri,walkin data entry operator night shift ites jo...
1,Work Based Onhome Based Part Time.,find live infotech,Chennai,Marketing,Job Description Send me Jobs like this Quali...,0 - 0 yrs,naukri,work based onhome based part time. marketing j...
2,Pl/sql Developer - SQL,Softtech Career Infosystem Pvt. Ltd,Bengaluru,IT Software - Application Programming,Job Description Send me Jobs like this - as ...,4 - 8 yrs,naukri,pl sql developer sql it software application p...
3,Manager/ad/partner - Indirect Tax - CA,Onboard HRServices LLP,"Mumbai, Bengaluru, Kolkata, Chennai, Coimbator...",Accounts,Job Description Send me Jobs like this - Inv...,11 - 15 yrs,naukri,manager ad partner indirect tax ca accounts jo...
4,JAVA Technical Lead (6-8 yrs) -,Spire Technologies and Solutions Pvt. Ltd.,Bengaluru,IT Software - Application Programming,Job Description Send me Jobs like this Pleas...,6 - 8 yrs,naukri,java technical lead 6 8 yrs it software applic...


In [27]:
print("Job Corpus")
print("#"*30)
display(corpus_df.head())

Job Corpus
##############################


,title,company,location,skills,description,experience,source
0,Walkin Data Entry Operator (night Shift),MM Media Pvt Ltd,Chennai,ITES,Job Description Send me Jobs like this Quali...,0 - 1 yrs,naukri
1,Work Based Onhome Based Part Time.,find live infotech,Chennai,Marketing,Job Description Send me Jobs like this Quali...,0 - 0 yrs,naukri
2,Pl/sql Developer - SQL,Softtech Career Infosystem Pvt. Ltd,Bengaluru,IT Software - Application Programming,Job Description Send me Jobs like this - as ...,4 - 8 yrs,naukri
3,Manager/ad/partner - Indirect Tax - CA,Onboard HRServices LLP,"Mumbai, Bengaluru, Kolkata, Chennai, Coimbator...",Accounts,Job Description Send me Jobs like this - Inv...,11 - 15 yrs,naukri
4,JAVA Technical Lead (6-8 yrs) -,Spire Technologies and Solutions Pvt. Ltd.,Bengaluru,IT Software - Application Programming,Job Description Send me Jobs like this Pleas...,6 - 8 yrs,naukri


In [29]:
# Random Samples

print("Resume Dataset")
display(resume_df.sample(5, random_state=42))

print("Jobs Dataset")
display(jobs_df.sample(5, random_state=42))

print("Job Corpus")
display(corpus_df.sample(5, random_state=42))

Resume Dataset


,Category,text
344,Java Developer,technical skills programming languages java se...
381,Java Developer,technicalskills springmvc hibernate jdbc java ...
351,Java Developer,education details january 2013 master of engin...
778,Hadoop,skill set hadoop map reduce hdfs hive sqoop ja...
277,Health and fitness,education details january 1992 to january 2003...


Jobs Dataset


,title,company,location,skills,description,experience,source,text
124499,Scheduler - Transit and Rail,Mace,"New York, NY","Management, Manufacturing","At Mace, our purpose is to redefine the bounda...",NaN,linkedin,scheduler transit and rail management manufact...
44892,Maintenance Mechanic,CAPTEK® Softgel International,"Vista, CA","Engineering, Manufacturing, Other",Become part of a team that champions wellness ...,NaN,linkedin,maintenance mechanic engineering manufacturing...
95237,"Line Producer, CourtTV",The E.W. Scripps Company,"Atlanta, GA","Art/Creative, Design, Information Technology",CourtTV is currently looking for our next Line...,Mid-Senior level,linkedin,line producer courttv art creative design info...
124511,Digital Chief Compliance Officer,GE Vernova,"District of Columbia, United States","General Business, Human Resources, Legal",Job Description Summary\n\nThe Digital Chief C...,Director,linkedin,digital chief compliance officer general busin...
136638,Culinary Internship - Three Leaf Catering (Tem...,University of Notre Dame,"Notre Dame, IN","Education, Training",Job Description\n\nAre you passionate about co...,Internship,linkedin,culinary internship three leaf catering tempor...


Job Corpus


,title,company,location,skills,description,experience,source
124499,Scheduler - Transit and Rail,Mace,"New York, NY","Management, Manufacturing","At Mace, our purpose is to redefine the bounda...",NaN,linkedin
44892,Maintenance Mechanic,CAPTEK® Softgel International,"Vista, CA","Engineering, Manufacturing, Other",Become part of a team that champions wellness ...,NaN,linkedin
95237,"Line Producer, CourtTV",The E.W. Scripps Company,"Atlanta, GA","Art/Creative, Design, Information Technology",CourtTV is currently looking for our next Line...,Mid-Senior level,linkedin
124511,Digital Chief Compliance Officer,GE Vernova,"District of Columbia, United States","General Business, Human Resources, Legal",Job Description Summary\n\nThe Digital Chief C...,Director,linkedin
136638,Culinary Internship - Three Leaf Catering (Tem...,University of Notre Dame,"Notre Dame, IN","Education, Training",Job Description\n\nAre you passionate about co...,Internship,linkedin


### Observation

- Resume dataset contains 962 records.
- Job dataset contains 136,759 records.
- No dataset loading errors occurred.
- Column names are standardized.
- Data is ready for EDA.

In [32]:
validation_summary = pd.DataFrame({
    "Dataset": ["Resume", "Jobs", "Job Corpus"],
    "Rows": [
        resume_df.shape[0],
        jobs_df.shape[0],
        corpus_df.shape[0]
    ],
    "Columns": [
        resume_df.shape[1],
        jobs_df.shape[1],
        corpus_df.shape[1]
    ],
    "Missing Values": [
        resume_df.isnull().sum().sum(),
        jobs_df.isnull().sum().sum(),
        corpus_df.isnull().sum().sum()
    ],
    "Duplicate Rows": [
        resume_df.duplicated().sum(),
        jobs_df.duplicated().sum(),
        corpus_df.duplicated().sum()
    ]
})

display(validation_summary)

,Dataset,Rows,Columns,Missing Values,Duplicate Rows
0,Resume,962,2,0,796
1,Jobs,136759,8,31746,0
2,Job Corpus,136759,7,31746,0


## Conclusion

The processed datasets have been successfully validated.

Key findings:

- All datasets loaded successfully.
- Dataset dimensions are consistent.
- Expected columns are present.
- Missing values and duplicate records have been examined.
- The processed datasets are ready for Exploratory Data Analysis (EDA).